### PR TITLE
Additional data for registration

### PR DESCRIPTION
**Pull request by ankitecd**
Changes in this pull request:
- Added additional data forms and metadata to add data in each registration

Known bugs to solve:
- Form editor scrolls only when scrolled from edge of screen

### CODE REVIEW

**Comment by norbertschuler** [comment]
- [ ] Got this and an endless spinner after doing a single tree registration:

```
[Tue May 18 2021 18:27:42.462]  LOG      formdata=>>>>>>>>>>>>>>>> []
[Tue May 18 2021 18:27:42.540]  LOG      formdata=>>>>>>>>>>>>>>>> []
[Tue May 18 2021 18:27:42.631]  LOG      formdata=>>>>>>>>>>>>>>>> []
[Tue May 18 2021 18:27:42.635]  LOG      formdata=>>>>>>>>>>>>>>>> []
[Tue May 18 2021 18:27:42.640]  LOG      formdata=>>>>>>>>>>>>>>>> []
[Tue May 18 2021 18:27:42.640]  LOG      formdata=>>>>>>>>>>>>>>>> []
[Tue May 18 2021 18:27:42.676]  LOG      formdata=>>>>>>>>>>>>>>>> []
[Tue May 18 2021 18:27:42.726]  LOG      formdata=>>>>>>>>>>>>>>>> []
[Tue May 18 2021 18:27:42.728]  LOG      formdata=>>>>>>>>>>>>>>>> []
[Tue May 18 2021 18:27:42.729]  LOG      formdata=>>>>>>>>>>>>>>>> []
[Tue May 18 2021 18:27:42.733]  LOG      forms []
[Tue May 18 2021 18:27:42.894]  WARN     Possible Unhandled Promise Rejection (id: 0):
TypeError: undefined is not an object (evaluating 'formsData[index].title')

```
Simulator Screen Shot - iPhone 11 - 2021-05-18 at 18 28 16

**Comment by norbertschuler** [comment]
> Got this and an endless spinner after doing a single tree or multiple tree registration - maybe my TPO account is missing data or is my database scheme not correct (upgrade from a former test) or what am I missing?

- [ ] I think that maybe I have to build myself a form first (it should not crash for people not having build a form), so I am testing build forms, but for adding a first input element, I get this error:

```
Unrecognized font family 'UberMoveMedium'
+[RCTFont updateFont:withFamily:size:weight:style:variant:scaleMultiplier:]
    RCTFont.mm:329
```

File: app/components/AdditionalData/DeleteAction.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  add,
+  lessThan,
+  multiply,
+} from 'react-native-reanimated';
+import { Colors } from '../../styles';
+
+const styles = StyleSheet.create({
+  remove: {
+    color: 'white',
+    fontFamily: 'UberMoveMedium',
```
Comment (norbertschuler) [inline app/components/AdditionalData/DeleteAction.tsx:18]: This does not work on iOS - where does this font came from?

Comment (ankitecd) [reply app/components/AdditionalData/DeleteAction.tsx]: Fixed

**Comment by norbertschuler** [comment]
- [ ] Is there any UI concept how to present the meta data in the review screen to the user - this technical representation surely is not the final design ;-)

Simulator Screen Shot - iPhone 11 - 2021-05-27 at 09 54 07

**Comment by norbertschuler** [comment]
Some thoughts on additional data editor:
- [ ] The selection of registration and tree type to ask the additional data for is not visible afterwards any more (same as mandatory and public flag) and as the data field does not seem to be editable again (or have I missed the functionality?) one also cannot check the definition again.
- [ ] The screens to add an elements and defining it's name and details is missing some padding.
- [ ] Adding dropdown options delete all previously configured values for the dropdown again.
- [ ] Changing the registration type deletes/overrides any previous selection of the tree type again.
- [ ] I have no idea what Metadata are used for? 🤔

Simulator Screen Shot - iPhone 11 - 2021-05-27 at 09 56 03Simulator Screen Shot - iPhone 11 - 2021-05-27 at 09 56 16

**Comment by ankitecd** [comment]
> 
> 
>     * [ ]  Is there any UI concept how to present the meta data in the review screen to the user - this technical representation surely is not the final design ;-)
> 

yes this not final yet.

**Comment by ankitecd** [comment]
> 
>     * [ ]  I have no idea what Metadata are used for? 🤔
> 

Here you can add the data which will be included in all the tree registrations. This does not appear like a form while doing the a tree registration but is included in each registration. Basically global data for all the registrations.

**Comment by norbertschuler** [comment]
> > ```
> > * [ ]  I have no idea what Metadata are used for? 🤔
> > ```
> 
> Here you can add the data which will be included in all the tree registrations. This does not appear like a form while doing the a tree registration but is included in each registration. Basically global data for all the registrations.

Thank you for explaining. But I have not seen my defined metadata being added when doing some registrations yet.

**Comment by ankitecd** [comment]
> 
> 
> > > ```
> > > * [ ]  I have no idea what Metadata are used for? 🤔
> > > ```
> > 
> > 
> > Here you can add the data which will be included in all the tree registrations. This does not appear like a form while doing the a tree registration but is included in each registration. Basically global data for all the registrations.
> 
> Thank you for explaining. But I have not seen my defined metadata being added when doing some registrations yet.



> 
> 
> > > ```
> > > * [ ]  I have no idea what Metadata are used for? 🤔
> > > ```
> > 
> > 
> > Here you can add the data which will be included in all the tree registrations. This does not appear like a form while doing the a tree registration but is included in each registration. Basically global data for all the registrations.
> 
> Thank you for explaining. But I have not seen my defined metadata being added when doing some registrations yet.

Actually I forgot to push the code. Now its available. Also this commit solves the issue of editing the input when clicked on border.

**Comment by ankitecd** [comment]
> 
>     * [ ]  Are the additional data definitions stored anywhere outside the app, e.g. with API calls as the species, so they do not get lost if deleting the app?
> 

This is still in development. The definitions are not going to be stored on server. There will a feature to export and import the definitions in the app. Right now we are planning to have a QR code export for now but if it fails to store large data then we will shift to export and import of json file.

**Comment by norbertschuler** [comment]
- [ ] @ankitecd I got the strange issue that I can not edit the input fields for height and diameter or tag anymore on iOS? Can you reproduce this on Android?

Simulator Screen Shot - iPhone 11 - 2021-05-31 at 11 25 42

**Comment by ankitecd** [comment]
> 
> 
>     * [ ]  @ankitecd I got the strange issue that I can not edit the input fields for height and diameter or tag anymore on iOS? Can you reproduce this on Android?
> 

Fixed. Did a mistake while fixing other issue.

**Comment by norbertschuler** [comment]
- [ ] Shortly before rendering the review screen it already shows some undefined metadata. Maybe this can be resolved when replacing the current metadata display against the final design as mentioned in https://github.com/Plant-for-the-Planet-org/treemapper/pull/327#issuecomment-849423778 (this screenshot is showing the transition between the former screen and the review screen - please be not be disturbed by this): 

Simulator Screen Shot - iPhone 11 - 2021-05-31 at 15 09 31

**Comment by ankitecd** [comment]
> 
> 
>     * [ ]  Import & Export do not work on iOS (share screen closes again immediately after opening and import does not show anything at all):
> 

iCloud needs to be enabled as mentioned here
https://www.npmjs.com/package/react-native-document-picker

**Comment by norbertschuler** [comment]
> * [ ]  Import & Export do not work on iOS (share screen closes again immediately after opening and import does not show anything at all):

The fact that on iOS the share dialog is closed (or hidden) instantly again after having selected the menu item to export must be somehow related to calling the function from the menu. When I tested opening the export share dialog from a normal button which I placed after the header component, it was not hidden again.

File: app/components/AdditionalData/index.tsx [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+      />
+    ),
+    metadata: () => (
+      <Metadata
+        {...{ metadata, updateMetadataOrder, addMetadataInState, loading: metadataLoading }}
+      />
+    ),
+  });
+
+  const handleImportExport = async (option: OptionsType) => {
```
Comment (norbertschuler) [inline app/components/AdditionalData/index.tsx:174]: For Android we need to check for the right permission to be granted: https://github.com/Plant-for-the-Planet-org/treemapper/blob/8fe00c0183ca3f29e69257cfe86b5c005d8ba7d0/app/components/InventoryOverview/index.js#L281-L288

**Review by norbertschuler — CHANGES_REQUESTED** [review]

File: app/components/Common/MenuOptions/index.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+          </TouchableOpacity>
+        }>
+        {options.map((option: OptionsType) => (
+          <MenuItem
+            key={option.key}
+            disabled={option.disabled}
+            style={styles.option}
+            onPress={() => {
+              onOptionPress(option);
+              hideMenu();
```
Comment (norbertschuler) [inline app/components/Common/MenuOptions/index.tsx:59]: Once you hide the menu also all opened dialog, e.g. the share dialog, get hidden as well! How can we solve this?

**Review by norbertschuler — CHANGES_REQUESTED** [review]

File: app/components/Common/MenuOptions/index.tsx (hunk 3 above)
Comment (norbertschuler) [reply app/components/Common/MenuOptions/index.tsx:59]: Related to https://github.com/Plant-for-the-Planet-org/treemapper/pull/327#issuecomment-852849981 and https://github.com/Plant-for-the-Planet-org/treemapper/pull/327#issuecomment-853642036

Comment (norbertschuler) [reply app/components/Common/MenuOptions/index.tsx:59]: I just tried to trigger the `onOptionPress` from the `onHidden` callback in the menu component of https://www.npmjs.com/package/react-native-material-menu (storing the last selected menu items in a local state), but the problem persist that the share dialog gets hidden again :-(

**Comment by norbertschuler** [comment]
I made a test branch at https://github.com/Plant-for-the-Planet-org/treemapper/pull/360/files to demonstrate what I tried out.

**Comment by norbertschuler** [comment]
- [ ] Testing the upload of the metadata works, but the private data I defined within my app did not get posted to the backend - this was the JSON code included in the API call:
 ```
"metadata": {
    "public": {
      "Test Public": "2",
      "Test": "1"
    },
    "private": {},
    "app": {}
  },
```
And this was part of the registration within the app, so please test if private data maybe got forgotten?

Simulator Screen Shot - iPhone 11 - 2021-06-03 at 14 52 36

- [ ] What is supposed to be used posted with `"app": {}`?

File: app/components/AdditionalData/index.tsx (hunk 2 above)
Comment (norbertschuler) [reply app/components/AdditionalData/index.tsx:174]: This just needs to be check for the export and write permission, not import needing only read permission.

**Comment by norbertschuler** [comment]
- [ ] I do not see all additional data in the list of uploaded registrations if I look at my latest uploaded registrations (downloaded from the server after having deleted and installed the app) again - can it be that if I export the additional data definitions, delete the app and install it again and then import the definition file, it does not match with the additional data fields stored on the server e.g. the following?

```json
 {
    "id": "loc_T0Ay2wgLKUbVtCuq1y8XSjWE",
    "type": "single",
    "captureMode": "on-site",
    "captureStatus": "partial",
    "geometry": {
      "type": "Point",
      "coordinates": [
        9.935324799421949,
        53.54902274208003
      ]
    },
    "coordinates": [
      {
        "coordinateIndex": "0",
        "image": "60bf47d8ca6fd182102071.jpg",
        "status": "complete",
        "id": "coord_3sZaEDjed4kxULAEIeDU48t9"
      }
    ],
    "plantDate": "2021-06-08T00:00:00+00:00",
    "plantedSpecies": [],
    "deviceLocation": {
      "type": "Point",
      "coordinates": [
        9.935324799421949,
        53.54902274208003
      ]
    },
    "plantProject": null,
    "scientificSpecies": "sspec_i08DZUvNbyL0alNuEq",
    "plantProjectSite": null,
    "registrationDate": "2021-06-08T00:00:00+00:00",
    "measurements": {
      "height": 0.3048,
      "width": 2.54
    },
    "tag": null,
    "metadata": {
      "public": {
        "Test Public": "2",
        "Test": "1"
      },
      "private": {
        "INPUT-1622623929920": "3",
        "DROPDOWN-1623148356764": "2"
      },
      "app": []
    }
  }
```

Simulator Screen Shot - iPhone 11 - 2021-06-08 at 12 47 34Simulator Screen Shot - iPhone 11 - 2021-06-08 at 12 47 37Simulator Screen Shot - iPhone 11 - 2021-06-08 at 12 47 50

This is the content of the exported file:

Simulator Screen Shot - iPhone 11 - 2021-06-08 at 12 55 09

**Comment by ankitecd** [comment]
> 
> Shouldn't the app be able to display all kind of additional data structures for registrations independent from the currently loaded definitions (or maybe no definition is loaded at all)? It might be a common use case that a TPO is changing its additional data definitions over time, so having registrations with different additional data definitions.
> 
> Also how should we display the data later on our website where we do not have the definition file?
> 
> I think we need to build a general display engine for this data?

Yes I realized that right now metadata are being shown from the schema and not from the inventory. That's why it is showing the metadata which is currently present in the app regardless of whether the inventory have additional data or not.
Changing that implemetation to show metadata from the inventory and not from the schema, so metadata will not be visible for older version registrations also.

Also while saving the inventory the additional data are not being stored so including that also now. This will solve this issue

**Comment by norbertschuler** [comment]
- [ ] Why do we repeat certain registration data now in the app part of the metadata (repeating showing them in the review screen as well)? E.g.

```JSON
{
  "type": "single",
  "captureMode": "on-site",
  "deviceLocation": {
    "coordinates": [
      9.93549999999999,
      53.5485
    ],
    "type": "Point"
  },
  "geometry": {
    "coordinates": [
      9.93549999999999,
      53.5485
    ],
    "type": "Point"
  },
  "plantDate": "2021-06-10",
  "registrationDate": "2021-06-10",
  "metadata": {
    "public": {
      "Test20 Public": "20",
      "Test10": "10"
    },
    "private": {
      "INPUT-1622623929920": "5",
      "DROPDOWN-1623148356764": "1"
    },
    "app": {
      "specieHeight": "0.3048",
      "specieDiameter": "2.54",
      "tagId": "3434",
      "speciesId1": "sspec_i08DZUvNbyL0alNuEq",
      "speciesAliases1": "Arboa antsingyae",
      "speciesTreeCount1": "1"
    }
  },
  "measurements": {
    "height": 0.3048,
    "width": 2.54
  },
  "scientificSpecies": "sspec_i08DZUvNbyL0alNuEq",
  "tag": "3434"
}
```

Simulator Screen Shot - iPhone 11 - 2021-06-10 at 11 45 34

**Comment by ankitecd** [comment]
> 
>     * [ ]  Why do we repeat certain registration data now in the app part of the metadata (repeating showing them in the review screen as well)? E.g.
> 
Sagar has asked to keep it for now and same it available while exporting GeoJSON

**Review by norbertschuler — APPROVED** [review]

**Review by norbertschuler — APPROVED** [review]

**Pull request merged**